src/test/py/bazel/genrule_test.py

```diff
@@ -0,0 +1,148 @@
+
+    copied = os.path.join(bazel_bin, 'foo', 'hello_copied')
+    self.assertTrue(os.path.exists(copied))
+    self.AssertFileContentContains(copied, 'hello world')
+
+  def testCopyWithBashAndPowershell(self):
+    self.ScratchFile('WORKSPACE')
+    self.ScratchFile('foo/BUILD', [
+      'genrule(',
+      '  name = "x",',
+      '  srcs = ["hello"],',
+      '  outs = ["hello_copied"],',
+      '  cmd_bash = "cp $< $@",',
+      '  cmd_ps = "Copy-Item $< -Destination $@",',
+      ')',
+    ])
+    self.ScratchFile('foo/hello', ['hello world'])
+
+    exit_code, stdout, stderr = self.RunBazel(['info', 'bazel-bin'])
+    self.AssertExitCode(exit_code, 0, stderr)
+    bazel_bin = stdout[0]
+
+    exit_code, _, stderr = self.RunBazel(['build', '//foo:x'])
+    self.AssertExitCode(exit_code, 0, stderr)
+
+    copied = os.path.join(bazel_bin, 'foo', 'hello_copied')
+    self.assertTrue(os.path.exists(copied))
+    self.AssertFileContentContains(copied, "hello world")
+
+  def testScriptFileIsUsedWithBatch(self):
+    if not self.IsWindows():
+      return
+    self.ScratchFile('WORKSPACE')
+    self.ScratchFile('foo/BUILD', [
+      'genrule(',
+      '  name = "x",',
+      '  outs = ["hello_world"],',
+      '  cmd_bat = \'&& \'.join([\"echo Hello world>>$(location hello_world)\" for _ in range(0, 1000)]),',
+      ')',
+    ])
+
+    exit_code, stdout, stderr = self.RunBazel(['info', 'bazel-bin'])
+    self.AssertExitCode(exit_code, 0, stderr)
+    bazel_bin = stdout[0]
+
+    exit_code, _, stderr = self.RunBazel(['build', '//foo:x'])
+    self.AssertExitCode(exit_code, 0, stderr)
+
+    script = os.path.join(bazel_bin, 'foo', 'x.genrule_script.bat')
+    hello = os.path.join(bazel_bin, 'foo', 'hello_world')
+    self.assertTrue(os.path.exists(script))
+    self.assertTrue(os.path.exists(hello))
+
+    expected_content = '\n'.join(['Hello world' for _ in range(0, 1000)])
+    self.AssertFileContentContains(hello, expected_content)
+
+  def testScriptFileIsUsedWithPowershell(self):
+    if not self.IsWindows():
+      return
+    self.ScratchFile('WORKSPACE')
+    self.ScratchFile('foo/BUILD', [
+      'genrule(',
+      '  name = "x",',
+      '  outs = ["hello_world"],',
+      '  cmd_ps ="; ".join(["echo \\"Hello world\\">>$(location hello_world)" for _ in range(0, 1000)]),',
+      ')',
+    ])
+
+    exit_code, stdout, stderr = self.RunBazel(['info', 'bazel-bin'])
+    self.AssertExitCode(exit_code, 0, stderr)
+    bazel_bin = stdout[0]
+
+    exit_code, _, stderr = self.RunBazel(['build', '//foo:x'])
+    self.AssertExitCode(exit_code, 0, stderr)
+
+    script = os.path.join(bazel_bin, 'foo', 'x.genrule_script.ps1')
+    hello = os.path.join(bazel_bin, 'foo', 'hello_world')
+    self.assertTrue(os.path.exists(script))
+    self.assertTrue(os.path.exists(hello))
+
+    expected_content = '\n'.join(['Hello world' for _ in range(0, 1000)])
+    self.AssertFileContentContains(hello, expected_content)
```

Comment:
@laszlocsomor This will fail without `$PSDefaultParameterValues['*:Encoding'] = 'utf8'`, because we read the file content with the default encoding (utf-8). This is how I discovered Powershell is using utf-16 as default encoding. I think it does exact what you suggested in other comments?

Reply:
Yes. Thanks!